common/src/main/java/com/wynntils/core/features/FeatureBase.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.features;
+
+import com.google.common.base.CaseFormat;
+import com.google.common.collect.ImmutableList;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.mc.utils.keybinds.KeyHolder;
+import com.wynntils.mc.utils.keybinds.KeyManager;
+import java.util.ArrayList;
+import java.util.List;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TranslatableComponent;
+
+public abstract class FeatureBase extends Feature {
+    private boolean isListener = false;
+    private List<KeyHolder> keyMappings = new ArrayList<>();
+
+    /**
+     * Sets up this feature as an event listener. This should be called from the constructor.
+     * Behaviour if calling it after the feature is enabled is undefined.
+     */
+    public void setupEventListener() {
+        this.isListener = true;
+    }
+
+    /**
+     * Adds a keyHolder to the feature. This should be called from the constructor.
+     * Behaviour if calling it after the feature is enabled is undefined.
+     * @param keyHolder KeyHolder to add to the feature
+     */
+    public void setupKeyHolder(KeyHolder keyHolder) {
+        keyMappings.add(keyHolder);
+    }
```

Comment:
However, it isn't really a setup, but more of a registration. Maybe it be called `asEventListener` and `withKeybind`, respectively. On that note, it'd be nice to have something like

```java
    public void setupKeyHolders(KeyHolder... keyHolders) {
        keyMappings. addAll(Arrays.asList(keyHolders));
    }
```

Reply:
`withKeybind` feels like it's a FluentAPI thing though.

Reply:
I ran through a couple of design iterations on this. In the end, I chose the `setup` prefix to emphasize the common treats between the two. Basically, it's saying there's two "ways" a feature can interact with Minecraft, either from keypresses or from events. I can very much imagine we can expand this to even more ways in the future, paving the way for more `setup` functions.

Reply:
As for a method to add multiple keyholders, I actually decided against that. I think that it will be clearer if the constructors list all KeyHolders on separate lines. (And right now we don't have any multi key features at all.) If we end up with a feature adding like a dozen keys or so, we might reconsider it.

Reply:
mhm